File: cc/mac/aes_cmac_key_manager_test.cc
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////
#include "tink/mac/aes_cmac_key_manager.h"

#include <memory>
#include <string>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "tink/chunked_mac.h"
#include "tink/mac.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"
#include "tink/util/test_matchers.h"
#include "proto/aes_cmac.pb.h"

namespace crypto {
namespace tink {

namespace {

using ::crypto::tink::test::IsOk;
using ::google::crypto::tink::AesCmacKey;
using ::google::crypto::tink::AesCmacKeyFormat;
using ::google::crypto::tink::AesCmacParams;
using ::testing::Eq;
using ::testing::Not;
using ::testing::SizeIs;

TEST(AesCmacKeyManagerTest, Basics) {
  EXPECT_THAT(AesCmacKeyManager().get_version(), Eq(0));
  EXPECT_THAT(AesCmacKeyManager().get_key_type(),
              Eq("type.googleapis.com/google.crypto.tink.AesCmacKey"));
  EXPECT_THAT(AesCmacKeyManager().key_material_type(),
              Eq(google::crypto::tink::KeyData::SYMMETRIC));
}

TEST(AesCmacKeyManagerTest, ValidateEmptyKey) {
  EXPECT_THAT(AesCmacKeyManager().ValidateKey(AesCmacKey()), Not(IsOk()));
}

AesCmacParams ValidParams() {
  AesCmacParams params;
  params.set_tag_size(16);
  return params;
}

AesCmacKeyFormat ValidKeyFormat() {
  AesCmacKeyFormat format;
  *format.mutable_params() = ValidParams();
  format.set_key_size(32);
  return format;
}

TEST(AesCmacKeyManagerTest, ValidateEmptyKeyFormat) {
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(AesCmacKeyFormat()),
              Not(IsOk()));
}

TEST(AesCmacKeyManagerTest, ValidateSimpleKeyFormat) {
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(ValidKeyFormat()), IsOk());
}

TEST(AesCmacKeyManagerTest, ValidateKeyFormatKeySizes) {
  AesCmacKeyFormat format = ValidKeyFormat();

  format.set_key_size(0);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.set_key_size(1);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.set_key_size(15);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.set_key_size(16);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.set_key_size(17);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.set_key_size(31);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.set_key_size(32);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), IsOk());

  format.set_key_size(33);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));
}

TEST(AesCmacKeyManagerTest, ValidateKeyFormatTagSizes) {
  AesCmacKeyFormat format = ValidKeyFormat();

  format.mutable_params()->set_tag_size(0);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.mutable_params()->set_tag_size(9);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.mutable_params()->set_tag_size(10);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), IsOk());

  format.mutable_params()->set_tag_size(11);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), IsOk());

  format.mutable_params()->set_tag_size(12);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), IsOk());

  format.mutable_params()->set_tag_size(15);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), IsOk());

  format.mutable_params()->set_tag_size(16);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), IsOk());

  format.mutable_params()->set_tag_size(17);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));

  format.mutable_params()->set_tag_size(32);
  EXPECT_THAT(AesCmacKeyManager().ValidateKeyFormat(format), Not(IsOk()));
}

TEST(AesCmacKeyManagerTest, CreateKey) {
  AesCmacKeyFormat format = ValidKeyFormat();
  ASSERT_THAT(AesCmacKeyManager().CreateKey(format), IsOk());
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  EXPECT_THAT(key.version(), Eq(0));
  EXPECT_THAT(key.key_value(), SizeIs(format.key_size()));
  EXPECT_THAT(key.params().tag_size(), Eq(format.params().tag_size()));
}

TEST(AesCmacKeyManagerTest, ValidateKey) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  EXPECT_THAT(AesCmacKeyManager().ValidateKey(key), IsOk());
}

TEST(AesCmacKeyManagerTest, ValidateKeyInvalidVersion) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  key.set_version(1);
  EXPECT_THAT(AesCmacKeyManager().ValidateKey(key), Not(IsOk()));
}

TEST(AesCmacKeyManagerTest, ValidateKeyShortKey) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  key.set_key_value("0123456789abcdef");
  EXPECT_THAT(AesCmacKeyManager().ValidateKey(key), Not(IsOk()));
}

TEST(AesCmacKeyManagerTest, ValidateKeyLongTagSize) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  key.mutable_params()->set_tag_size(17);
  EXPECT_THAT(AesCmacKeyManager().ValidateKey(key), Not(IsOk()));
}


TEST(AesCmacKeyManagerTest, ValidateKeyTooShortTagSize) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  key.mutable_params()->set_tag_size(9);
  EXPECT_THAT(AesCmacKeyManager().ValidateKey(key), Not(IsOk()));
}

TEST(AesCmacKeyManagerTest, GetMacPrimitive) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();
  auto manager_mac_or = AesCmacKeyManager().GetPrimitive<Mac>(key);
  ASSERT_THAT(manager_mac_or, IsOk());
  auto mac_value_or = manager_mac_or.value()->ComputeMac("some plaintext");
  ASSERT_THAT(mac_value_or, IsOk());

  auto direct_mac_or = subtle::AesCmacBoringSsl::New(
      util::SecretDataFromStringView(key.key_value()), key.params().tag_size());
  ASSERT_THAT(direct_mac_or, IsOk());
  EXPECT_THAT(
      direct_mac_or.value()->VerifyMac(mac_value_or.value(), "some plaintext"),
      IsOk());
}

TEST(AesCmacKeyManagerTest, GetChunkedMacPrimitive) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();

  util::StatusOr<std::unique_ptr<ChunkedMac>> chunked_mac =
      AesCmacKeyManager().GetPrimitive<ChunkedMac>(key);
  ASSERT_THAT(chunked_mac, IsOk());

  util::StatusOr<std::unique_ptr<ChunkedMacComputation>> computation =
      (*chunked_mac)->CreateComputation();
  ASSERT_THAT(computation, IsOk());
  ASSERT_THAT((*computation)->Update("abc"), IsOk());
  ASSERT_THAT((*computation)->Update("xyz"), IsOk());
  util::StatusOr<std::string> tag = (*computation)->ComputeMac();
  ASSERT_THAT(tag, IsOk());

  util::StatusOr<std::unique_ptr<ChunkedMacVerification>> verification =
      (*chunked_mac)->CreateVerification(*tag);
  ASSERT_THAT(verification, IsOk());
  ASSERT_THAT((*verification)->Update("abc"), IsOk());
  ASSERT_THAT((*verification)->Update("xyz"), IsOk());
  EXPECT_THAT((*verification)->VerifyMac(), IsOk());
}

TEST(AesCmacKeyManagerTest, MixPrimitives) {
  AesCmacKeyFormat format = ValidKeyFormat();
  AesCmacKey key = AesCmacKeyManager().CreateKey(format).value();

  util::StatusOr<std::unique_ptr<Mac>> mac =
      AesCmacKeyManager().GetPrimitive<Mac>(key);
  ASSERT_THAT(mac, IsOk());

  util::StatusOr<std::unique_ptr<ChunkedMac>> chunked_mac =
      AesCmacKeyManager().GetPrimitive<ChunkedMac>(key);
  ASSERT_THAT(chunked_mac, IsOk());

  // Compute tag with Mac.
  util::StatusOr<std::string> tag = (*mac)->ComputeMac("abcxyz");
  ASSERT_THAT(tag, IsOk());

  // Compute chunked tag with ChunkedMac.
  util::StatusOr<std::unique_ptr<ChunkedMacComputation>> computation =
      (*chunked_mac)->CreateComputation();
  ASSERT_THAT(computation, IsOk());
  ASSERT_THAT((*computation)->Update("abc"), IsOk());
  ASSERT_THAT((*computation)->Update("xyz"), IsOk());
  util::StatusOr<std::string> chunked_tag = (*computation)->ComputeMac();
  ASSERT_THAT(chunked_tag, IsOk());
  ASSERT_THAT(*chunked_tag, Eq(*tag));  // Both primitives generated same tag.

  // Verify chunked tag with Mac.
  ASSERT_THAT((*mac)->VerifyMac(*chunked_tag, "abcxyz"), IsOk());

  // Verify tag with ChunkedMac.
  util::StatusOr<std::unique_ptr<ChunkedMacVerification>> verification =
      (*chunked_mac)->CreateVerification(*tag);
  ASSERT_THAT(verification, IsOk());
  ASSERT_THAT((*verification)->Update("abc"), IsOk());
  ASSERT_THAT((*verification)->Update("xyz"), IsOk());
  EXPECT_THAT((*verification)->VerifyMac(), IsOk());
}

}  // namespace
}  // namespace tink
}  // namespace crypto

File: java/src/test/java/com/google/crypto/tink/mac/AesCmacKeyManagerTest.java
// Copyright 2017 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.mac;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.fail;

import com.google.crypto.tink.Mac;
import com.google.crypto.tink.proto.AesCmacKey;
import com.google.crypto.tink.proto.AesCmacKeyFormat;
import com.google.crypto.tink.proto.AesCmacParams;
import com.google.crypto.tink.subtle.AesCmac;
import com.google.crypto.tink.subtle.Random;
import com.google.protobuf.ByteString;
import java.security.GeneralSecurityException;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Test for AesCmacKeyManager. */
@RunWith(JUnit4.class)
public class AesCmacKeyManagerTest {
  @Test
  public void validateKeyFormat_empty() throws Exception {
    try {
      new AesCmacKeyManager().keyFactory().validateKeyFormat(AesCmacKeyFormat.getDefaultInstance());
      fail();
    } catch (GeneralSecurityException e) {
      // expected.
    }
  }

  private static AesCmacKeyFormat makeAesCmacKeyFormat(int keySize, int tagSize) {
    return AesCmacKeyFormat.newBuilder()
        .setKeySize(keySize)
        .setParams(AesCmacParams.newBuilder().setTagSize(tagSize).build())
        .build();
  }

  @Test
  public void validateKeyFormat_valid() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 10));
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 11));
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 12));
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 13));
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 14));
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 15));
    manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 16));
  }

  @Test
  public void validateKeyFormat_notValid_throws() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    try {
      manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 9));
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 17));
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(32, 32));
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(16, 16));
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager.keyFactory().validateKeyFormat(makeAesCmacKeyFormat(64, 16));
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
  }

  @Test
  public void createKey_valid() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 16)));
  }

  @Test
  public void createKey_checkValues() throws Exception {
    AesCmacKeyFormat keyFormat = makeAesCmacKeyFormat(32, 16);
    AesCmacKey key = new AesCmacKeyManager().keyFactory().createKey(keyFormat);
    assertThat(key.getKeyValue()).hasSize(keyFormat.getKeySize());
    assertThat(key.getParams().getTagSize()).isEqualTo(keyFormat.getParams().getTagSize());
  }

  @Test
  public void createKey_multipleTimes() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    AesCmacKeyFormat keyFormat = makeAesCmacKeyFormat(32, 16);
    assertThat(manager.keyFactory().createKey(keyFormat).getKeyValue())
        .isNotEqualTo(manager.keyFactory().createKey(keyFormat).getKeyValue());
  }

  @Test
  public void validateKey_valid() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 10)));
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 11)));
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 12)));
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 13)));
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 14)));
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 15)));
    manager.validateKey(manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 16)));
  }

  @Test
  public void validateKey_wrongVersion_throws() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    try {
      AesCmacKey validKey = manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 16));
      manager.validateKey(AesCmacKey.newBuilder(validKey).setVersion(1).build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
  }

  @Test
  public void validateKey_notValid_throws() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    AesCmacKey validKey = manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 16));
    try {
      manager.validateKey(
          AesCmacKey.newBuilder(validKey)
              .setKeyValue(ByteString.copyFrom(Random.randBytes(16)))
              .build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager.validateKey(
          AesCmacKey.newBuilder(validKey)
              .setKeyValue(ByteString.copyFrom(Random.randBytes(64)))
              .build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager
          .validateKey(
              AesCmacKey.newBuilder(validKey)
                  .setParams(AesCmacParams.newBuilder(validKey.getParams()).setTagSize(0).build())
                  .build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager
          .validateKey(
              AesCmacKey.newBuilder(validKey)
                  .setParams(AesCmacParams.newBuilder(validKey.getParams()).setTagSize(9).build())
                  .build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager
          .validateKey(
              AesCmacKey.newBuilder(validKey)
                  .setParams(AesCmacParams.newBuilder(validKey.getParams()).setTagSize(17).build())
                  .build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
    try {
      manager
          .validateKey(
              AesCmacKey.newBuilder(validKey)
                  .setParams(AesCmacParams.newBuilder(validKey.getParams()).setTagSize(32).build())
                  .build());
      fail();
    } catch (GeneralSecurityException e) {
      // expected
    }
  }


  @Test
  public void getPrimitive_works() throws Exception {
    AesCmacKeyManager manager = new AesCmacKeyManager();
    AesCmacKey validKey = manager.keyFactory().createKey(makeAesCmacKeyFormat(32, 16));
    Mac managerMac = manager.getPrimitive(validKey, Mac.class);
    Mac directMac =
        new AesCmac(validKey.getKeyValue().toByteArray(), validKey.getParams().getTagSize());
    byte[] message = Random.randBytes(50);
    managerMac.verifyMac(directMac.computeMac(message), message);
  }
}
